packages/@aws-cdk/cli-lib/lib/cli.ts

```diff
@@ -0,0 +1,276 @@
+   */
+  destroy(options?: DestroyOptions): Promise<void>;
+}
+
+/**
+ * Configuration for creating a CLI from an AWS CDK App directory
+ */
+export interface CdkAppDirectoryProps {
+  /**
+   * Command-line for executing your app or a cloud assembly directory
+   * e.g. "node bin/my-app.js"
+   * or
+   * "cdk.out"
+   *
+   * @default - read from cdk.json
+   */
+  readonly app?: string
+
+  /**
+   * Emits the synthesized cloud assembly into a directory
+   *
+   * @default cdk.out
+   */
+  readonly output?: string;
+}
+
+/**
+ * A class returning the path to a Cloud Assembly Directory when its `produce` method is invoked with the current context
+ *
+ * AWS CDK apps might need to be synthesized multiple times with additional context values before they are ready.
+ * When running the CLI from inside a directory, this is implemented by invoking the app multiple times.
+ * Here the `produce()` method provides this multi-pass ability.
+ */
+export interface ICloudAssemblyDirectoryProducer {
+  /**
+   * The working directory used to run the Cloud Assembly from.
+   * This is were a `cdk.context.json` files will be written.
+   *
+   * @default - current working directory
+   */
+  workingDirectory?: string;
+
+  /**
+   * Synthesize a Cloud Assembly directory for a given context.
+   *
+   * For all features to work correctly, `cdk.App()` must be instantiated with the received context values in the method body.
+   * Usually obtained similar to this:
+   * ```ts fixture=imports
+   * class MyProducer implements ICloudAssemblyDirectoryProducer {
+   *   async produce(context: Record<string, any>) {
+   *     const app = new cdk.App({ context });
+   *     // create stacks here
+   *     return app.synth().directory;
+   *   }
+   * }
+   * ```
+   */
+  produce(context: Record<string, any>): Promise<string>
+}
+
+/**
+ * Provides a programmatic interface for interacting with the AWS CDK CLI
+ */
+export class AwsCdkCli implements IAwsCdkCli {
+  /**
+   * Create the CLI from a directory containing an AWS CDK app
+   * @param directory the directory of the AWS CDK app. Defaults to the current working directory.
+   * @param props additional configuration properties
+   * @returns an instance of `AwsCdkCli`
+   */
+  public static fromCdkAppDirectory(directory?: string, props: CdkAppDirectoryProps = {}) {
+    return new AwsCdkCli(async (args) => {
+      const originalWorkingDir = process.cwd();
+      if (directory) {
+        process.chdir(directory);
+      }
+
+      if (props.app) {
+        args.push('--app', props.app);
+      }
+      if (props.output) {
+        args.push('--output', props.output);
+      }
+
+      const result = await runCli(args);
+
+      if (directory) {
+        process.chdir(originalWorkingDir);
+      }
+
+      return result;
+    });
+  }
+
+  /**
+   * Create the CLI from a CloudAssemblyDirectoryProducer
+   */
+  public static fromCloudAssemblyDirectoryProducer(producer: ICloudAssemblyDirectoryProducer) {
```

Comment:
Did we drop the `fromCloudAssemblyDirectory` constructor on purpose?

Reply:
Yes, I feel there's no point to it. If someone _really_ wants to do this, they could just do a stub producer:
```ts
{
  produce: async () => '/path/to/assembly/dir'
}
```

Reply:
Ya but it looks nicer

Reply:
But it's useless! We can always add it later if people are asking for it. 🤷🏻‍♂️